### PR TITLE
drivers: spi: Add basic support rpi-pico SPI (PrimeCell SSP PL022)

### PR DESCRIPTION
**Pull request by soburi**
Add fundamental feature support for rpi-pico's SPI peripheral.
This commit implements synchronous transfer with 8bit-MSB format.

I verified my environment with two test programs.
tests/drivers/spi/spi_loopback test is work well.
(rpi_pico seems to have a problem when running device-testing  by the twister,
 locally build and flashing and check console output.)
And also samples/subsys/fs/fat_fs/ works.

rp2040 datasheet says the SoC has ARM PrimeCell SSP PL022 as the SPI peripheral.
So this PL022 driver implementation can compile without rpi_pico SDK.

Optional feature are not currently implemented yet.
It is work in future.

- DMA transfer is not implemented yet.
- Slave mode is not implemented yet.
- Currently support only 8-bit data transfer.

Hardware limitation:

- LSB-first format is not supported by hardware.

### CODE REVIEW

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi
Comment (yonsch) [inline boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:37]: TX and SCK should be in their own group, where input is not enabled

File: boards/arm/rpi_pico/rpi_pico.yaml
Comment (yonsch) [inline boards/arm/rpi_pico/rpi_pico.yaml:15]: Add SPI to the "Supported Features" table in `boards/arm/rpi_pico/doc.index.rst` also`

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (yonsch) [inline dts/arm/rpi_pico/rp2040.dtsi:104]: SPI should also be a reset device. @andrei-edward-popa , I suppose the correct way to do this is to create a separate DTS binding for rpi_pico, and have it inherit pl022 and reset-device?

File: include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h
Comment (yonsch) [inline include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h:112]: This should get up to pin 29, see table on page 13 of the RP2040 datasheet

**Review by yonsch — CHANGES_REQUESTED** [review]
My review for the Pico parts

**Comment by yonsch** [comment]
Two more things:
1. Why is CSN missing from the board's pinmux?
2. Set a default clock-frequency in `rpi_pico.dts`

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi
Comment (soburi) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: Fix it.

File: boards/arm/rpi_pico/rpi_pico.yaml
Comment (soburi) [reply boards/arm/rpi_pico/rpi_pico.yaml:15]: Fixed it.

File: include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h
Comment (soburi) [reply include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h:112]: Fixed it.

**Comment by soburi** [comment]
Hi, @yonsch,

> * Why is CSN missing from the board's pinmux?

It is my mistakes. Just forgot it.
CS pin is often controlled by 'cs-gpios' as gpio, so I misunderstood that I didn't must need it. 

> * Set a default clock-frequency in `rpi_pico.dts`

Set 8MHz to default. 
It can works also in peripheral mode (pclk/12 =~ 10MHz) and with enough safety margin.

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (andrei-edward-popa) [reply dts/arm/rpi_pico/rp2040.dtsi:104]: @soburi @yonsch Yes, you can create a DTS bindings under dts/bindings/spi/raspberrypi,pico-spi.yaml and inherit arm,pl022.yaml and reset,device.yaml (see raspberrypi,pico-uart.yaml). After that, add resets field inside spi nodes (take a look inside rpi_pico_common.dtsi for spi reset values).

File: drivers/spi/spi_pl022.c
Comment (yonsch) [inline drivers/spi/spi_pl022.c:386]: Declare those variables at the beginning of the function

**Review by yonsch — CHANGES_REQUESTED** [review]
Thanks, it looks better. I added a few comments about the PL022 driver, though it should be more thoroughly reviewed by someone with more expertise in this field.

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (yonsch) [reply dts/arm/rpi_pico/rp2040.dtsi]: Right, you don't need another implementation. Just make sure that enabling SPI in DTS still selects the PL022 implementation

File: drivers/spi/spi_pl022.c
Comment (soburi) [reply drivers/spi/spi_pl022.c]: Fixed it.

File: drivers/spi/spi_pl022.c
Comment (soburi) [reply drivers/spi/spi_pl022.c:285]: Fixed it.

File: boards/arm/rpi_pico/doc/index.rst
Comment (soburi) [reply boards/arm/rpi_pico/doc/index.rst]: Fixed it.

File: drivers/spi/spi_pl022.c
Comment (soburi) [reply drivers/spi/spi_pl022.c]: Fixed it.
Rewrite internally used defines.

File: include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h
Comment (soburi) [reply include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h]: Oh.. It's my careless.
Fixed it.

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (andrei-edward-popa) [reply dts/arm/rpi_pico/rp2040.dtsi]: I think you can remove arm,pl022 from the compatible strings because raspberrypi,pico-spi yaml inherits arm,pl022.yaml.

Comment (soburi) [reply dts/arm/rpi_pico/rp2040.dtsi]: > I think you can remove arm,pl022 from the compatible string because raspberrypi,pico-spi yaml inherits arm,pl022.yaml.

If do so, It couldn't resolve pl022 in DT_INST_FOREACH_STATUS_OKAY, can't compile it.
It seem to need it.

Compiling with CONFIG_SPI_PL022, I get such as,

```
/home/crs/zephyrproject/zephyr/drivers/spi/spi_pl022.c:473:12: warning: 'spi_pl022_init' defined but not used [-Wunused-function]
  473 | static int spi_pl022_init(const struct device *dev)
      |            ^~~~~~~~~~~~~~
/home/crs/zephyrproject/zephyr/drivers/spi/spi_pl022.c:468:30: warning: 'spi_pl022_api' defined but not used [-Wunused-variable]
  468 | static struct spi_driver_api spi_pl022_api = {
      |                              ^~~~~~~~~~~~~
```

Comment (andrei-edward-popa) [reply dts/arm/rpi_pico/rp2040.dtsi]: Okay, so probably that's the right way.

**Review by yonsch — DISMISSED** [review]
LGTM, Thanks!

**Comment by yonsch** [comment]
@soburi To comply with #45410, please change includes to `#include <zephyr/..>` where applicable.

**Comment by soburi** [comment]
@yonsch 

> @soburi To comply with #45410, please change includes to `#include <zephyr/..>` where applicable.

Thank you for notify.
I updated it.

**Review by mbolivar-nordic — COMMENTED** [review]
Devicetree portions all LGTM. @tbursztyka could you please take a look at this one?

**Comment by petejohanson** [comment]
@soburi Sadly, conflicts after the USB device driver info got added to the docs file. Needs a rebase.

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi
Comment (petejohanson) [inline boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:45]: According to the [Pico Datasheet](https://datasheets.raspberrypi.com/pico/Pico-R3-A4-Pinout.pdf) the only "default" SPI is SPI0, I would say we shouldn't enable/default `spi1` defaults at all for the pico board.

File: boards/arm/rpi_pico/rpi_pico.dts
Comment (petejohanson) [inline boards/arm/rpi_pico/rpi_pico.dts:97]: Per my other comment, I think we shouldn't be setting up `spi1` at all with defaults, leave that to end users.

**Review by petejohanson — COMMENTED** [review]
One minor comment on the spi1 defaults for the Pico board definition.

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi
Comment (soburi) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: Rebasing and Removed spi1 and related definitions.

Question: i2c1 is also not set as default in this illustration. Is it have no problem?
(I made SPI1 by seeing that.)

File: boards/arm/rpi_pico/rpi_pico.dts
Comment (soburi) [reply boards/arm/rpi_pico/rpi_pico.dts]: ditto.

**Comment by soburi** [comment]
@petejohanson ,

> @soburi Sadly, conflicts after the USB device driver info got added to the docs file. Needs a rebase.

I rebased to newest.

File: boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi
Comment (petejohanson) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: I didn't have a chance to review the i2c PR, but I would suggest we remove that for consistency in a separate PR.

Comment (soburi) [reply boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi]: Thanks for reply. I understand and agree it.

**Review by yonsch — DISMISSED** [review]

File: drivers/spi/spi_pl022.c
Comment (tbursztyka) [inline drivers/spi/spi_pl022.c:448]: Is there any possibility to get this driver working through interrupts instead of polling?

**Review by tbursztyka — DISMISSED** [review]

File: drivers/spi/spi_pl022.c
Comment (soburi) [reply drivers/spi/spi_pl022.c:448]: > Is there any possibility to get this driver working through interrupts instead of polling?

Currently, only work with polling. Not handles interrupt.
But I think it should also support the CONFIG_SPI_ASYNC option by handling interrupts
I have a plan to implement an interrupts handler later.

**Review by mbolivar-nordic — DISMISSED** [review]
Thank you for the nice work on the devicetree and driver model parts, +1 for these.

**Review by yonsch — DISMISSED** [review]

**Comment by yonsch** [comment]
@tbursztyka Could you give this another look?

**Comment by yonsch** [comment]
@soburi Please rebase and fix conflicts

**Comment by soburi** [comment]
@yonsch 

> @soburi Please rebase and fix conflicts

Thanks for notifying me.

I rebased it to the newest source.
I found a failure of tests/drivers/spi/spi_loopback test with the newest source.
And I fixed it and also updated this source.

**Review by yonsch — DISMISSED** [review]

**Comment by carlescufi** [comment]
@petejohanson  and @andrei-edward-popa could you please take another look?

**Review by petejohanson — DISMISSED** [review]
Tested here on a XIAO RP2040 controller on a keyboard w/ a shift register on the SPI bus, working perfectly w/ defaults, and also if I enable the interrupt mode.

Docs, DT bits, etc all look good. My high level understanding of the SPI driver pieces I don't see any major issues.

**Comment by andrei-edward-popa** [comment]
Looks great to me! I cannot wait to try the SPI in a real application with Pico. Good job!

**Review by petejohanson — CHANGES_REQUESTED** [review]
Code changes tested here, and working fine w/ my 595 shift register driver, and sharp memory display.

The compliance check is failing right now though, sadly, due to the commit message of the last commit being just "overlay".

**Comment by soburi** [comment]
Hi, @petejohanson 
> Code changes tested here, and working fine w/ my 595 shift register driver, and sharp memory display.
> 
> The compliance check is failing right now though, sadly, due to the commit message of the last commit being just "overlay".

Very sorry. I accidentally pushed the code I put in for confirmation with it committed. Deleted the last commit.

**Review by petejohanson — APPROVED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
DT bits look OK to me. @tbursztyka please take another look

**Comment by yonsch** [comment]
@soburi Please check if this compiles when `CONFIG_LEGACY_INCLUDE_PATH=n`. If it doesn't, you should add a few `zephyr/` prefixes to your includes. One of these instances was fixed in a recent PR, so you will have to rebase first. Other than that, LGTM.

**Comment by soburi** [comment]
@yonsch 
> @soburi Please check if this compiles when `CONFIG_LEGACY_INCLUDE_PATH=n`. If it doesn't, you should add a few `zephyr/` prefixes to your includes. One of these instances was fixed in a recent PR, so you will have to rebase first. Other than that, LGTM.

Thanks for notification.
I verified it, and found no problems.

**Review by petejohanson — DISMISSED** [review]

**Review by petejohanson — DISMISSED** [review]

**Comment by soburi** [comment]
- Use DT_HAS_ARM_PL022_ENABLED instead of dt_compat_enabled in Kconfig.pl022
- Call spi_context_complete if SPI_ASYNC enabled even if not enabled SPI_PL022_INTERRUPT.

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (galak) [inline dts/arm/rpi_pico/rp2040.dtsi:113]: remove `label`,  not needed

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (galak) [inline dts/arm/rpi_pico/rp2040.dtsi:126]: remove label

File: tests/drivers/spi/spi_loopback/boards/rpi_pico.conf
Comment (galak) [inline tests/drivers/spi/spi_loopback/boards/rpi_pico.conf:8]: Seen reworked PR for spi_loopback: https://github.com/zephyrproject-rtos/zephyr/pull/48222

Comment (soburi) [reply tests/drivers/spi/spi_loopback/boards/rpi_pico.conf:8]: Hi, @galak, 

I'll prepare to apply the #48222 change to this PR.

I also failed #48222 same as @erwango observed.

@erwango mentioned in #48222

> @galak, as expected this test failed on our bench since #48358 was merged. Out of curiosity I launched a campaign using current PR and I still see issues so most likely it requires a bit more work. Do you have the same status on your side, or do you want me to dig a little bit more ?

I'm checking #48222 progress, and I'll report when the update is complete.

Thank you.

Comment (soburi) [reply tests/drivers/spi/spi_loopback/boards/rpi_pico.conf:8]: Hi, @galak 

#48222 is correctly worked with also rpi_pico now.
I prepare to update this PR.

**Comment by soburi** [comment]
Update done. This PR should merge after #48222 merged.

File: dts/bindings/spi/arm,pl022.yaml
Comment (galak) [inline dts/bindings/spi/arm,pl022.yaml:17]: `interrupt` is `false` by default so not needed.

Comment (soburi) [reply dts/bindings/spi/arm,pl022.yaml]: Removed it.

File: tests/drivers/spi/spi_loopback/boards/rpi_pico.conf
Comment (soburi) [reply tests/drivers/spi/spi_loopback/boards/rpi_pico.conf]: Rebase done.

**Comment by soburi** [comment]
CI is unstable at this time. The fails seems not relate this PR.
I'll rebase if need.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by yonsch — APPROVED** [review]

**Comment by yonsch** [comment]
@soburi Congratulations, and thank you for your patience!

**Comment by soburi** [comment]
Hi, @yonsch ,

thank you for your helps.
And also please #45137, It makes life happy with GDB!

**Pull request merged**
